### ISSUE TITLE
Synonyms don't work correctly for Sensors (and possibly other tables?)

### ISSUE THREAD

**Issue by robintw** [debrief/pepys-import#387]
As quoted from Ian in #379,

"GPS-B" may be the synonym for "GPS-2" on "HMS DRAGON". But, "GPS-B" may be the synonym for another sensor on another platform. So, maybe there are some tables we can't support synonyms for. I'll have to give this more thought.

This means we could run into issues using synonyms at the moment, as they may end up referring to the wrong sensor on the wrong platform.

**Comment by robintw** [comment]
The most frequently used table that this will affect is the Sensors table, as described above - a sensor is uniquely identified partially by the platform it is attached to, so that would need to be taken into account in the Synonyms table.

Synonyms also don't really make sense for certain tables, mostly because they're internal tables that the user doesn't directly change anyway:

 - Change
 - Log
 - Extraction (unused presently)
 - Tag (unused presently)
 - TaggedItem (unused presently)
 - HostedBy - this doesn't really have a 'name' to have a synonym of, so can't really work with synonyms anyway (unused presently)

Currently, synonym searching is implemented for Sensors, Platforms and Datafiles. A question for @IanMayo: are synonyms actually going to be used for all three of those? Do Platforms need to have synonyms? I can't quite think of how a Datafile needs a synonym, but I may be wrong. If there are objects that don't need to support synonyms then we can remove support and make the code easier to understand and maintain.

However, one of the most useful places to have synonyms seems to be with Sensor objects - and that is the bit that is currently impossible. The way to solve this problem on how critical the synonym functionality is for various tables. Options include:

**1. Don't allow synonym use for tables where it doesn't work properly.**
This can be done by changing the `add_to_synonyms` function to fail if it gets given certain table names - and then modifying the command-line resolver to not try to create synonyms for those particular tables. This would reduce the usefulness of synonyms - particularly for sensors which may be referred to by different names in different situations. In most importers we are now caching the sensor used by a specific platform, so users would only have to select the correct sensor once for each file - but that still adds to the user interaction requirements, and we're trying to reduce those.

**2. Add some extra columns to the Synonyms table to deal with other filters we need to have alongside the Synonym**
At the moment we search for a synonym by just looking in the Synonyms table for an entry with a given name, and extracting the GUID of the object that name refers to. We could extend the Synonyms table with two optional fields: `filter_column` and `filter_value`. When storing a synonym for a Sensor, the `filter_column` would be set to `host` and the `filter_value` would be set to the platform_id for the host of that sensor. When searching for a Synonym, if those fields have values in them, they will be added to the filters in the query - and that will mean that we only return the right Sensor object.
This would mean that we could have multiple synonyms with the same name, pointing to different Sensor objects, as long as they each had different entries for filter_value. This would allow `GPS-B` to point to Sensor 5, which belongs to Platform 1, but also `GPS-B` to point to Sensor 7 which belongs to Platform 2.

What do you think @IanMayo?

Also, as a note to myself: there is currently an extra bug in line 428 of command_line_resolver.py, where the wrong sensor may be added to the synonym in the first place.

**Comment by IanMayo** [comment]
1.  Agreed - I'd be ok with just supporting a limited list of synonyms:
* Sensor (Sonar 2088 could have: S2088, 2088, 2888R1).  So, that's not just `GPS-B` on multiple platforms, it's multiple ways of naming one sonar.  Aah, that's a new way of looking at a Sensor synonyms.  Parser has retrieved platform, then we try to retrieve sensor, and don't find it.  We then loop through the sensor synonyms and see if one matches. Hmm, our data structure doesn't support this.    I'm tempted to not offer synonyms for sensor
* Platform (maybe include common typos/ shortenings `RCHMND` for Richmond)
* Nationality (potentially things like: United Kingdom, GBR, GB)
* Media Types (CD could have synonyms `CDR`, `CD-ROM`)

Hmm, the more I look at the `Reference` block of tables, the more they look like candidates for synonyms.

So, (**180 degree turnaround**) as you suggested, we would actually black-list tables that don't support synonyms.  For now, I'm happy to add `Sensors` and `GeometrySubTypes` for that. Well, actually it would be the _secondary_ tables - ones that have to connect to a parent table.  Maybe we can get by with this restricted subset of tables.  If not, then I propose we consider alternate synonym strategies, that respond to the specific scenario the analysts are experiencing.

So, in summary, I'm proposing that we don't allow Sensor synonyms.  We'll push forward with synonyms for the other tables, then come back to sensors if/when we need to.

**Comment by robintw** [comment]
Decision based on discussions on Slack:

1. Blacklist tables that *cannot* have Synonyms - including `Sensors` and `GeometrySubTypes`

2. Remove synonym support in command-line resolver and in functions for searching, for both `Sensors` and `Datafiles`.
